examples/arkode/C_serial/ark_anharmonic_symplectic.c

```diff
@@ -0,0 +1,307 @@
+#include <sundials/sundials_types.h>
+
+#include "arkode/arkode.h"
+
+typedef struct
+{
+  int order;
+  int num_output_times;
+  int use_compsums;
+  sunrealtype dt;
+} ProgramArgs;
+
+/* RHS functions */
+static int pdot(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data);
+static int qdot(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data);
+
+/* Helper functions */
+static sunrealtype Hamiltonian(N_Vector yvec, sunrealtype t);
+static int ParseArgs(int argc, char* argv[], ProgramArgs* args);
+static void PrintHelp();
+static int check_retval(void* returnvalue, const char* funcname, int opt);
+
+int main(int argc, char* argv[])
+{
+  ProgramArgs args;
+  SUNContext sunctx  = NULL;
+  N_Vector y         = NULL;
+  sunrealtype* ydata = NULL;
+  sunrealtype tout   = NAN;
+  sunrealtype tret   = NAN;
+  void* arkode_mem   = NULL;
+  int iout           = 0;
+  int retval         = 0;
+
+  /* Parse the command line arguments */
+  if (ParseArgs(argc, argv, &args)) { return 1; };
+
+  /* Default integrator options */
+  int order                  = args.order;
+  int use_compsums           = args.use_compsums;
+  const int num_output_times = args.num_output_times;
+
+  /* Default problem parameters */
+  const sunrealtype T0    = SUN_RCONST(0.0);
+  sunrealtype Tf          = SUN_RCONST(30.0);
+  sunrealtype dt          = args.dt;
+  const sunrealtype dTout = (Tf - T0) / ((sunrealtype)num_output_times);
+
+  /* Create the SUNDIALS context object for this simulation */
+  retval = SUNContext_Create(NULL, &sunctx);
+  if (check_retval(&retval, "SUNContext_Create", 1)) return 1;
+
+  printf("\n   Begin time-dependent anharmonic oscillator problem\n\n");
+
+  /* Allocate our state vector */
+  y = N_VNew_Serial(4, sunctx);
+
+  /* Fill the initial conditions */
+  ydata    = N_VGetArrayPointer(y);
+  ydata[0] = 0; /* ptau */
+  ydata[1] = 0; /* \dot{q} = p */
+  ydata[2] = 0; /* qtau */
+  ydata[3] = 1; /* \ddot{q} = \dot{p} */
+
+  /* Create SPRKStep integrator */
+  arkode_mem = SPRKStepCreate(qdot, pdot, T0, y, sunctx);
+
+  retval = SPRKStepSetOrder(arkode_mem, order);
+  if (check_retval(&retval, "SPRKStepSetOrder", 1)) { return 1; }
+
+  retval = SPRKStepSetUseCompensatedSums(arkode_mem, use_compsums);
+  if (check_retval(&retval, "SPRKStepSetUseCompensatedSums", 1)) { return 1; }
+
+  retval = SPRKStepSetFixedStep(arkode_mem, dt);
+  if (check_retval(&retval, "SPRKStepSetFixedStep", 1)) { return 1; }
+
+  retval = SPRKStepSetMaxNumSteps(arkode_mem, ((long int)ceil(Tf / dt)) + 2);
+  if (check_retval(&retval, "SPRKStepSetMaxNumSteps", 1)) { return 1; }
+
+  /* Print out starting Hamiltonian before integrating */
+  tret = T0;
+  tout = T0 + dTout;
+  /* Output current integration status */
+  fprintf(stdout, "t = %.6Lf, q(t) = %.6Lf, H = %.6Lf\n", tret,
+          ydata[1], Hamiltonian(y, tret));
+
+  /* Do integration */
+  for (iout = 0; iout < num_output_times; iout++)
+  {
+    SPRKStepSetStopTime(arkode_mem, tout);
```

Comment:
It might be nice if the program included a command-line argument to enable/disable "tstop" mode (like you do for compensated sums), so that it would be easy to test the effects of interpolation.  However, this could be a "someday to-do" and shouldn't hold up the PR.

Reply:
Added in each of the examples